app/frontend/wallet/cardano-wallet.ts

```diff
@@ -191,99 +248,57 @@ const CardanoWallet = async (options) => {
-        txInputs,
-        address,
-        (reducedAmount + diff) as Lovelace,
-        roundedDonation
-      )
-      return {
-        sendAmount: reducedAmount + diff - txFee,
-        donationAmount: roundedDonation,
-      }
+      const txFee = computeRequiredTxFee(inputs, outputs)
+      return {sendAmount: Math.max(coins - txFee, 0)}
     } else {
-      const txFee = computeTxFee(
-        txInputs,
-        address,
-        (coins - donationAmount) as Lovelace,
-        donationAmount
-      )
-      return {sendAmount: Math.max(coins - donationAmount - txFee, 0)}
+      const inputs = profitableUtxos
+      const outputs = [{address, coins: 0}, {address: ADA_DONATION_ADDRESS, coins: 0}]
+      const txFee = computeRequiredTxFee(inputs, outputs)
+
+      if (donationType === 'percentage') {
+        // set maxSendAmount and percentageDonation (0.2% of max) to deplete balance completely
+        const percent = 0.2
+
+        const reducedAmount: Lovelace = Math.floor(coins / (1 + percent / 100)) as Lovelace
+        const roundedDonation = roundWholeAdas(((reducedAmount * percent) / 100) as Lovelace)
+
+        return {
+          sendAmount: coins - txFee - roundedDonation,
+          donationAmount: roundedDonation,
+        }
+      } else {
+        return {sendAmount: Math.max(coins - donationAmount - txFee, 0)}
+      }
     }
   }
 
-  async function getMaxSendableAmount(address, hasDonation, donationAmount, donationType) {
-    const {txInputs, coins} = await getTxInputsAndCoins()
-
-    if (!hasDonation) {
-      const txFee = computeTxFee(txInputs, address, coins, false)
-      return {sendAmount: Math.max(coins - txFee, 0)}
-    }
+  async function getMaxDonationAmount(address, sendAmount: Lovelace) {
+    const utxos = await getUTxOs()
+    const profitableUtxos = utxos.filter(isUtxoProfitable)
+    const coins = getAvailableBalance(profitableUtxos)
 
-    return getMaxSendableAmountWithDonation(address, txInputs, coins, donationAmount, donationType)
-  }
+    const inputs = profitableUtxos
+    const outputs = [{address, coins: 0}, {address: ADA_DONATION_ADDRESS, coins: 0}]
 
-  async function getMaxDonationAmount(address, sendAmount) {
-    const {txInputs, coins} = await getTxInputsAndCoins()
-    const txFee = computeTxFee(txInputs, address, sendAmount, true, coins - sendAmount)
+    const txFee = computeRequiredTxFee(inputs, outputs)
     return Math.max(coins - txFee - sendAmount, 0)
   }
 
-  async function getTxFee(address, coins, hasDonation, donationAmount) {
-    const txInputs = await prepareTxInputs(address, coins, hasDonation, donationAmount)
-    return computeTxFee(txInputs, address, coins, hasDonation, donationAmount)
+  async function getTxPlan(address, coins: Lovelace, donationAmount: Lovelace) {
+    const allUtxos = await getUTxOs()
+    const change = await getChangeAddress() // TODO: should be cached
+    const plan = selectMinimalTxPlan(allUtxos, address, coins, donationAmount, change)
```

Comment:
I would rename this const to `changeAddress` since its passed to a function that reffers to it as `changeAddress` and refers to `change` as address and coins.